psydac/api/ast/nodes.py

```diff
@@ -111,16 +159,25 @@ class IndexDofTrial(IndexDof):
 class IndexDofTest(IndexDof):
     pass
 
+class IndexOuterDofTest(IndexDof):
+    pass
+
+class IndexInnerDofTest(IndexDof):
+    pass
+
 class IndexDerivative(IndexNode):
     def __new__(cls, length=None):
         return Basic.__new__(cls)
 
-index_element   = IndexElement()
-index_quad      = IndexQuadrature()
-index_dof       = IndexDof()
-index_dof_test  = IndexDofTest()
-index_dof_trial = IndexDofTrial()
-index_deriv     = IndexDerivative()
+index_element        = IndexElement()
+index_quad           = IndexQuadrature()
+index_dof            = IndexDof()
+index_dof_test       = IndexDofTest()
+index_dof_trial      = IndexDofTrial()
+index_deriv          = IndexDerivative()
+index_outer_dof_test = IndexOuterDofTest()
+index_inner_dof_test = IndexInnerDofTest()
```

Comment:
Are these singletons? Could you explain why they are needed and how they are used?

Reply:
those represent the indices used in the loops and the arrays, and yes they are singletons.